lib/sda/treenode.go

```diff
@@ -0,0 +1,507 @@
+	// instance
+	msgQueue map[network.MessageTypeID][]*Data
+	// done callback
+	onDoneCallback func() bool
+	// queue holding msgs
+	msgDispatchQueue []*Data
+	// locking for msgqueue
+	msgDispatchQueueMutex sync.Mutex
+	// kicking off new message
+	msgDispatchQueueWait chan bool
+	// whether this node is closing
+	closing bool
+}
+
+// aggregateMessages (if set) tells to aggregate messages from all children
+// before sending to the (parent) Node
+// https://golang.org/ref/spec#Iota
+const (
+	aggregateMessages = 1 << iota
+)
+
+// MsgHandler is called upon reception of a certain message-type
+type MsgHandler func([]*interface{})
+
+// NewNode creates a new node
+func newTreeNodeInstance(o *Overlay, tok *Token, tn *TreeNode) *TreeNodeInstance {
+	n := &TreeNodeInstance{overlay: o,
+		token:                tok,
+		channels:             make(map[network.MessageTypeID]interface{}),
+		handlers:             make(map[network.MessageTypeID]interface{}),
+		messageTypeFlags:     make(map[network.MessageTypeID]uint32),
+		msgQueue:             make(map[network.MessageTypeID][]*Data),
+		treeNode:             tn,
+		msgDispatchQueue:     make([]*Data, 0, 1),
+		msgDispatchQueueWait: make(chan bool, 1),
+	}
+	go n.dispatchMsgReader()
+	return n
+}
+
+// TreeNode gets the treeNode of this node. If there is no TreeNode for the
+// Token of this node, the function will return nil
+func (n *TreeNodeInstance) TreeNode() *TreeNode {
+	return n.treeNode
+}
+
+// Entity returns our entity
+func (n *TreeNodeInstance) Entity() *network.Entity {
+	return n.treeNode.Entity
+}
+
+// Parent returns the parent-TreeNode of ourselves
+func (n *TreeNodeInstance) Parent() *TreeNode {
+	return n.treeNode.Parent
+}
+
+// Children returns the children of ourselves
+func (n *TreeNodeInstance) Children() []*TreeNode {
+	return n.treeNode.Children
+}
+
+// Root returns the root-node of that tree
+func (n *TreeNodeInstance) Root() *TreeNode {
+	return n.Tree().Root
+}
+
+// IsRoot returns whether whether we are at the top of the tree
+func (n *TreeNodeInstance) IsRoot() bool {
+	return n.treeNode.Parent == nil
+}
+
+// IsLeaf returns whether whether we are at the bottom of the tree
+func (n *TreeNodeInstance) IsLeaf() bool {
+	return len(n.treeNode.Children) == 0
+}
+
+// SendTo sends to a given node
+func (n *TreeNodeInstance) SendTo(to *TreeNode, msg interface{}) error {
+	if to == nil {
+		return errors.New("Sent to a nil TreeNode")
+	}
+	return n.overlay.SendToTreeNode(n.token, to, msg)
+}
+
+// Tree returns the tree of that node
+func (n *TreeNodeInstance) Tree() *Tree {
+	return n.overlay.TreeFromToken(n.token)
+}
+
+// EntityList returns the entity-list
+func (n *TreeNodeInstance) EntityList() *EntityList {
+	return n.Tree().EntityList
+}
+
+// Suite can be used to get the current abstract.Suite (currently hardcoded into
+// the network library). Preferably use this function instead of network.Suite
+// when possible.
```

Comment:
Not have precise plan in mind no. But I would have said the contrary ^^ I would prefer to see the code using functions that returns the suite used by the struct, and at the lowest level of the function stack, this `Suite` is defined to be the `network.Suite`. When we want to change, we have to change it only in one place. 
But since we don't really know what we want to do later (support multiple `Suite` ?), I don't know what would be the best solution ..